torch/_refs/__init__.py

```diff
@@ -3194,7 +3194,10 @@ def index_select(x: TensorLike, dim: int, index: TensorLike):
     )
     # Treat scalars as elements of \R^1
     if x.ndim == 0:
-        return x.unsqueeze(0)[index].squeeze(0).clone()
+        # we cannot write `x.unsqueeze(0)[index].squeeze(0).clone()`
+        # as tensor[index] will trigger index.item() if index is a 0-dim tensor
+        # and .item() cannot be symbolically traced with FakeTensor.
+        return torch.ops.aten.index(x.unsqueeze(0), [index]).squeeze(0).clone()
     idx = (slice(None),) * dim + (index,)
```

Comment:
I see we rewrote slice(None) into None in Mario's PR. Is it also needed here?

Reply:
Yes, it's needed.

Reply:
Yes it's needed, I don't understand why it's not caught with tests, cc @lezcano

Reply:
I tried replacing this with `(None, ) * dim + (index, )` and the decomp result is incorrect... 

With ` (slice(None),) * dim + (index,)`, index_select is decomposed into 
```
    def forward(self, primals_1, primals_2):
         = torch.ops.aten.slice.Tensor(primals_1, 0, 0, 9223372036854775807);  primals_1 = None
        index = torch.ops.aten.index.Tensor(slice_1, [None, primals_2]);  slice_1 = None
        return [index, primals_2]
```

With `(None, ) * dim + (index, )`, 

```
    def forward(self, primals_1, primals_2):
        unsqueeze = torch.ops.aten.unsqueeze.default(primals_1, 0);  primals_1 = None
        index = torch.ops.aten.index.Tensor(unsqueeze, [None, primals_2]);  unsqueeze = None
        return [index, primals_2]
```

Reply:
I think this is related to how indexing is interpreted with python API. 
x[slice(None), index] ==  x[:, index]  
x[None, index] == x.unsqueeze(0)[index]

Reply:
Ah ok, that's the difference between `x[idx]` syntax (that requires `slice`) and `torch.ops.aten.index.Tensor` syntax that doesn't.

Reply:
OK, this is really confusing... 


```
x = torch.arange(6).reshape(2, 3)
index = torch.tensor([0])

# Following lines works, and are equivalent
print(torch.ops.aten.index_select(x, 1, index))
print(x[slice(None), index])
print(x[:, index])
print(torch.ops.aten.index(x, [None, index]))

# This doesn't work !!!
print(torch.ops.aten.index(x, [slice(None), index]))
```

Reply:
Note that we need the `slice(None)` in the Python API, while we need `None` in the decomposition... Same happens for all the other `index_*` functions :/

Reply:
Sorry, no, her eit should be `slice(None)`. It should be `None` if you call into `aten.index_put_`, and `slice(None)` if you call into `__getitem__`.

Reply:
For reference, the issue is that `index` and these functions take a list of optional tensors, with `c10::nullopt` being equal to `:` (which in Python is equal to `slice(None)`). Now, `__getitem__` accepts many other things. It can accept a tuple of tensors, but also an `...` or an `Ellipsis`. All these Python constructions are not handled in ATen native, but at a lower level, before calling into ATen functions. These constructions are transformed into a list of optional Tensors which is then passed to `aten.index`.

A similar (but even more intricate) issue happens with `__setitem__`, constructions like `x[idx] = t` or `x[idx] += t`, and `{aten,torch}.index_put_(idx, accumulate=True)`

These functions have lots of legacy code which has not been maintained, so that's why we are where we are, sadly